### DISCUSSION TITLE
Heterogeneous Workers

### DISCUSSION THREAD

**Discussion by nviets** [wlandau/crew#23] [Q&A]
Will crew enable heterogeneous workers? Say, for instance, I have a modeling workflow with one task that needs to run on a GPU instance and a bunch of ETL and reporting tasks that can run on generic CPU nodes. Clustermq/targets don't appear to support this.

Snakemake supports per rule configuration of slurm partitions, which is a nice feature for specifying dedicated workers for a particular set of rules/targets.

**Comment by wlandau** [comment]
I do plan to handle this with {rrq}’s built-in ability to create and manage multiple queues. Different queues can have their own workers with their own resource requirements.

**Reply by wlandau** [reply]
With the `mirai` backend, this is going to be possible with multi-controllers: https://github.com/wlandau/crew/blob/main/vignettes/multi_controller.Rmd. The same mechanism will work with the `rrq` backend too, but since `rrq` already supports multiple queues, I doubt this will be necessary.

**Reply by nviets** [reply]
Great work, @wlandau. How do you envision this feature working with targets? Will there be a tar_make_crew to supersede tar_make_clustermq? I was imagining something like the below, but let me know if I'm not thinking about this correctly:

```
list(
  tar_target(file, "data.csv", format = "file"),
  tar_target(data, get_data(file)),
  tar_target(model, fit_model_w_gpu(data), worker = "transient_gpu_worker"),
  tar_target(plot, plot_model(model, data))
)
```

**Reply by wlandau** [reply]
It's definitely going to be in a new `tar_make_crew()` function or just part of `tar_make()` (I haven't decided yet). Then in `_targets.R`, the user will create the `crew` controller/multicontroller globally. In the case of a multicontroller, you will be able to optionally pass a character vector of controller names to each target that has specialized requirements. (Again, I haven't decided on the interface to accept controller names. It will be either part of `tar_resources()` somehow or possibly a new argument to tar_target()`.)
